CRM/Contribute/BAO/ContributionRecur.php

```diff
@@ -856,13 +859,16 @@ public static function updateOnNewPayment($recurringContributionID, $paymentStat
 
     if (!empty($existing['installments']) && self::isComplete($recurringContributionID, $existing['installments'])) {
       $params['contribution_status_id'] = 'Completed';
+      $params['next_sched_contribution_date'] = '';
     }
     else {
-      // Only update next sched date if it's empty or 'just now' because payment processors may be managing
-      // the scheduled date themselves as core did not previously provide any help.
-      if (empty($existing['next_sched_contribution_date']) || strtotime($existing['next_sched_contribution_date']) ==
-        strtotime($effectiveDate)) {
-        $params['next_sched_contribution_date'] = date('Y-m-d', strtotime('+' . $existing['frequency_interval'] . ' ' . $existing['frequency_unit'], strtotime($effectiveDate)));
+      // Update next_sched_contribution_date in the following circumstances:
+      // - If it's empty.
+      // - If it is older than the contribution receive date.
+      // If we update, set it to the contribution receive date + recur frequency unit * interval
```

Comment:
We should probably note why we don't always do it - ie precautionary in case some other process has already handled it